### PR TITLE
Add lock to BaseRobot to prevent multiple robot processes.

### LINKED ISSUE #3: [FEATURE] Locks
We need to have a lock to prevent multiple robot processes. Suggestions are to use `/var/run`, but that does lock it down to be platform specific which we do not want.

### PR DESCRIPTION
**Pull request by trickeydan**
Resolves #3 

In previous APIs, this has been implemented using a lock file in `/tmp`. For this version it was suggested that we put a PID file in `/var/run`, which was an approach that I initially approved of.

However, up to this point and as confirmed by #120, we are ensuring the cross-platform compatibility of `j5` and I'm sure you're aware that Windows doesn't have `/var/run`. The closest equivalent is called a Mutex and has very poor support on Unix systems.

The solution that I have implemented in this pull request takes leafs out of the book of enterprise Java and isn't exactly what I would do in an ideal world.

Binding to a TCP port on the host is kept to a single process at the kernel level, which means that it is harder to override (e.g no deleting of the PID file.). Additionally, in the case of a crash, the TCP port is automatically closed as we are no longer binding to it.

I am still open to alternative solutions to this and am looking forward to hearing back from you all.

### CODE REVIEW

File: tests/test_stack.py [hunk 1]
```diff
@@ -19,3 +22,13 @@ def test_led():
 
     for led in r.demo_board.leds:
         led.state = True
+
+
+def test_robot_lock():
+    """Test that we cannot have more than one Robot object."""
+    r1 = Robot()
+
+    assert r1._lock
```
Comment (RealOrangeOne) [inline tests/test_stack.py]: What exactly is this asserting. It would be better to assert that the address bound to is what you were expecting.

**Comment by RealOrangeOne** [comment]
Whilst I also think that using a socket for this kind of thing is really weird, i've spent 20 minutes looking around for a cross-platform alternative, and came up empty. Although, almost any other application which needs an inter-process lock uses PID files, and I suspect there's a good reason for that! 

I'd also suggest using a much higher, less _l33t_ port. `j5` in ascii is `10653`, which would be a much better fitting port. 

(Also I agree with @sedders123 that locking should not be a public method)

File: tests/test_stack.py (hunk 1 above)
Comment (trickeydan) [reply tests/test_stack.py]: Fair point. Significantly improved.

**Comment by kierdavis** [comment]
I like this.

**Pull request merged**